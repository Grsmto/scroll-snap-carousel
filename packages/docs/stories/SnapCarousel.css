.snap-carousel-container {
  height: 300px;
}

.slide {
  /* width: calc(33.33% - ((2 * 10vw) / 3)); */
  width: 100%;
  flex-shrink: 0;
  height: 100%;
  scroll-snap-align: center;
  background-color: #c9edf2;
  background-clip: content-box;
  box-sizing: border-box;
  display: grid;
  place-items: center;
  padding: 0 15px;
}

.thumbnails .slide {
  width: calc(33.33% - ((2 * 10vw) / 3));
  scroll-snap-align: start;
}

/* .slide:first-child {
  margin-left: 10vw;
}

.slide:last-child {
  margin-right: 10vw;
} */

@media (max-width: 640px) {
  .slide {
    width: calc(50% - ((2 * 5vw) / 2));
  }

  .slide:first-child {
    margin-left: 5vw;
  }

  .slide:last-child {
    margin-right: 5vw;
  }
}

@media (max-width: 320px) {
  .slide {
    width: 100%;
  }

  .slide:first-child {
    margin-left: 20px;
  }

  .slide:last-child {
    margin-right: 20px;
  }
}

.carousel-indicator {
  display: flex;
  justify-content: center;
}

.carousel-indicator__dot {
  width: 10px;
  height: 10px;
  border-radius: 10px;
  background-color: lightgray;
}

.carousel-indicator__dot+.carousel-indicator__dot {
  margin-left: 10px;
}

.carousel-indicator__dot.active {
  background-color: lightgreen;
}
