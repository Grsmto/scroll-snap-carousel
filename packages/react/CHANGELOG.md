# v0.0.17 (Mon Dec 13 2021)

#### üêõ Bug Fix

- Bump independent versions \[skip ci\] ([@Grsmto](https://github.com/Grsmto))
- Update CHANGELOG.md \[skip ci\] ([@Grsmto](https://github.com/Grsmto))

#### ‚ö†Ô∏è Pushed to `master`

- update react dep version ([@Grsmto](https://github.com/Grsmto))
- bump versions ([@Grsmto](https://github.com/Grsmto))
- fix(react): fix not building ([@Grsmto](https://github.com/Grsmto))
- chore: rename package to @snap-carousel ([@Grsmto](https://github.com/Grsmto))

#### Authors: 1

- Adrien Denat ([@Grsmto](https://github.com/Grsmto))

---

# v0.0.16 (Sat Jan 02 2021)

#### üêõ Bug Fix

- Bump independent versions \[skip ci\] ([@Grsmto](https://github.com/Grsmto))
- Update CHANGELOG.md \[skip ci\] ([@Grsmto](https://github.com/Grsmto))

#### ‚ö†Ô∏è Pushed to `master`

- update react dep version ([@Grsmto](https://github.com/Grsmto))
- bump versions ([@Grsmto](https://github.com/Grsmto))
- fix(react): fix not building ([@Grsmto](https://github.com/Grsmto))
- chore: rename package to @snap-carousel ([@Grsmto](https://github.com/Grsmto))

#### Authors: 1

- Adrien Denat ([@Grsmto](https://github.com/Grsmto))

---

# v0.0.14 (Thu Dec 24 2020)

#### ‚ö†Ô∏è Pushed to `master`

- fix(react): fix not building ([@Grsmto](https://github.com/Grsmto))
- chore: rename package to @snap-carousel ([@Grsmto](https://github.com/Grsmto))

#### Authors: 1

- Adrien Denat ([@Grsmto](https://github.com/Grsmto))
