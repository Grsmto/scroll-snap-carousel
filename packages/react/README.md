# Snap Carousel React

### Installation

**- Via npm**
`npm install @snap-carousel/react --save`

**- Via Yarn**
`yarn add @snap-carousel/react`

### Usage

## Component

```js
import {
  SnapCarousel
} from '@snap-carousel/react';
import '@snap-carousel/react/dist/styles.css';

const Component = () => {
  let slides = [...];

  return (
    <SnapCarousel>
      {slides.map((slide) => (
        <div className="slide" key={slide}>
          {slide}
        </div>
      ))}
    </SnapCarousel>
  );
};
```

## Hooks

```js
import {
  useDragToScroll,
  useScroll,
  useActiveSnap,
} from '@snap-carousel/react';
import '@snap-carousel/core/dist/styles.css';

const Component = () => {
  const ref = React.useRef();

  useDragToScroll({ ref });

  let slides = [...];

  return (
    <div ref={ref}>
      {slides.map((slide) => (
        <div key={slide}>
          {slide}
        </div>
      ))}
    </div>
  );
};
```

For more advanced usage see [the Storybook stories](https://github.com/Grsmto/scroll-snap-carousel/blob/master/packages/docs/stories).

**Don't forget to import both css and js in your project!**
