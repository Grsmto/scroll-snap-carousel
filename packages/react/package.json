{
  "name": "@snap-carousel/react",
  "version": "0.0.21-next.3",
  "main": "dist/index.cjs",
  "module": "dist/index.js",
  "jsnext:main": "dist/index.js",
  "types": "dist/index.d.ts",
  "type": "module",
  "license": "MIT",
  "private": false,
  "repository": "https://github.com/Grsmto/scroll-snap-carousel",
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "build": "tsup",
    "start": "tsup --watch",
    "storybook": "storybook dev --port 6006",
    "build-storybook": "storybook build",
    "test-storybook": "test-storybook"
  },
  "dependencies": {
    "@snap-carousel/core": "workspace:*"
  },
  "peerDependencies": {
    "react": ">=16.8.0",
    "react-dom": ">=16.8.0"
  },
  "devDependencies": {
    "@storybook/addon-actions": "^8.4.7",
    "@storybook/addon-essentials": "^8.4.7",
    "@storybook/addon-interactions": "^8.4.7",
    "@storybook/addon-links": "^8.4.7",
    "@storybook/jest": "^0.2.3",
    "@storybook/react": "^8.4.7",
    "@storybook/react-vite": "^8.4.7",
    "@storybook/test-runner": "0.20.1",
    "@storybook/testing-library": "^0.2.2",
    "@testing-library/jest-dom": "^5.16.5",
    "@testing-library/react": "^13.4.0",
    "@types/react": "^18.0.17",
    "@types/react-dom": "^18.0.6",
    "@typescript-eslint/eslint-plugin": "^3.2.0",
    "@typescript-eslint/parser": "^3.2.0",
    "react": "^18.0.0",
    "react-dom": "^18.0.0",
    "storybook": "^8.4.7",
    "tsup": "^8.3.5"
  }
}
