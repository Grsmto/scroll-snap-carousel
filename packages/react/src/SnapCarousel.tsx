import React, { useCallback, useRef } from 'react';
import type { ElementType } from 'react';

import { useActiveSnap } from './useActiveSnap';
import { useDragToScroll } from './useDragToScroll';
import { useScroll } from './useScroll';
import { PolymorphicRef, useMergedRefs } from './utils';

type State = ReturnType<typeof useSnapCarousel>;

interface Props<T extends ElementType> {
  children?: React.ReactNode;
  className?: string;
  tag?: T;
  index?: number;
  defaultIndex?: number;
  onIndexChange?: (index: number) => void;
  onScrollEnd?: () => void;
  state?: State;
}

export const useSnapCarousel = () => {
  const elRef = useRef<any>(null);
  const activeIndex = useActiveSnap({ ref: elRef });

  return {
    index: activeIndex,
    elRef,
  };
};

export const SnapCarousel = React.forwardRef(
  <T extends ElementType = 'div'>(
    {
      children,
      tag,
      index,
      defaultIndex,
      onIndexChange,
      onScrollEnd,
      className,
      state,
      ...otherProps
    }: Props<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const RootTag = tag || 'div';
    const elRef = state?.elRef || useRef<any>(null);
    const mergedRef = useMergedRefs([elRef, ref]);
    const scrollTo = useScroll({ ref: elRef, onScrollEnd });
    const activeIndex = useActiveSnap({ ref: elRef });

    useDragToScroll({ ref: elRef });

    React.useEffect(() => {
      if (typeof index !== 'undefined') scrollTo(index);
    }, [index]);

    React.useEffect(() => {
      if (typeof defaultIndex !== 'undefined') scrollTo(defaultIndex, 0);
    }, [defaultIndex]);

    React.useEffect(() => {
      onIndexChange && onIndexChange(activeIndex);
    }, [activeIndex, onIndexChange]);

    return (
      <RootTag
        className={`snap-carousel-container ${className}`}
        ref={mergedRef}
        {...otherProps}
      >
        {children}
      </RootTag>
    );
  }
);

SnapCarousel.displayName = 'SnapCarousel';

type CommonProps<T extends ElementType> = {
  children?: React.ReactNode;
  className?: string;
  tag?: T;
  activeIndex?: number;
  state?: State;
};

interface IndicatorProps<T extends ElementType> extends CommonProps<T> {
  index: number;
}

export const SnapCarouselIndicator = React.forwardRef(
  <T extends ElementType = 'button'>(
    {
      tag,
      index,
      activeIndex,
      className = '',
      state,
      ...otherProps
    }: IndicatorProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const RootTag = tag || 'button';
    const elRef = state?.elRef || useRef<any>(null);
    const scrollTo = useScroll({ ref: elRef });
    const _activeIndex =
      typeof activeIndex === 'number'
        ? activeIndex
        : typeof state?.index === 'number'
        ? state.index
        : undefined;

    const handleClick = useCallback(() => {
      scrollTo(index);
    }, [index]);

    return (
      <RootTag
        className={`${index === _activeIndex ? ' active' : ''}${
          className ? ' ' + className : ''
        }`}
        ref={ref}
        onClick={handleClick}
        {...otherProps}
      />
    );
  }
);

export const SnapCarouselNavPrev = React.forwardRef(
  <T extends ElementType = 'button'>(
    {
      tag,
      children,
      activeIndex,
      className = '',
      state,
      ...otherProps
    }: CommonProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const RootTag = tag || 'button';
    const elRef = state?.elRef || useRef<any>(null);
    const scrollTo = useScroll({ ref: elRef });
    const _activeIndex =
      typeof activeIndex === 'number'
        ? activeIndex
        : typeof state?.index === 'number'
        ? state.index
        : undefined;

    const handleClick = useCallback(() => {
      if (typeof _activeIndex === 'undefined') return;
      scrollTo(_activeIndex - 1);
    }, [_activeIndex]);

    return (
      <RootTag
        className={`${className ? ' ' + className : ''}`}
        ref={ref}
        onClick={handleClick}
        disabled={_activeIndex === 0}
        aria-hidden={_activeIndex === 0}
        aria-label="Previous"
        {...otherProps}
      >
        {children}
      </RootTag>
    );
  }
);

export const SnapCarouselNavNext = React.forwardRef(
  <T extends ElementType = 'button'>(
    {
      tag,
      children,
      activeIndex,
      className = '',
      state,
      ...otherProps
    }: CommonProps<T>,
    ref?: PolymorphicRef<T>
  ) => {
    const RootTag = tag || 'button';
    const elRef = state?.elRef || useRef<any>(null);
    const scrollTo = useScroll({ ref: elRef });
    const _activeIndex =
      typeof activeIndex === 'number'
        ? activeIndex
        : typeof state?.index === 'number'
        ? state.index
        : undefined;

    const handleClick = useCallback(() => {
      if (typeof _activeIndex === 'undefined') return;
      scrollTo(_activeIndex + 1);
    }, [_activeIndex]);

    const length = elRef.current?.children.length;

    return (
      <RootTag
        className={`${className ? ' ' + className : ''}`}
        ref={ref}
        onClick={handleClick}
        disabled={_activeIndex === length - 1}
        aria-hidden={_activeIndex === length - 1}
        aria-label="Next"
        {...otherProps}
      >
        {children}
      </RootTag>
    );
  }
);
