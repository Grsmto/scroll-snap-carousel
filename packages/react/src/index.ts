import { utils } from '@snap-carousel/core';

export const { isTouchDevice } = utils;

// export { SnapItem, SnapList } from './SnapList';
export { useDragToScroll } from './useDragToScroll';
export { useScroll } from './useScroll';
export { useActiveSnap } from './useActiveSnap';
// export { useVisibleElements } from './useVisibleElements';
