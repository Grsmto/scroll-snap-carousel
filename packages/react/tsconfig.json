{
  "extends": "../../tsconfig.json",
  "compilerOptions": {
    "removeComments": false,
    "sourceMap": true,
    "inlineSources": true,
    "jsx": "react",
    "outDir": "./dist",
    "declaration": true,
    "declarationDir": "./dist"
  },
  "include": ["src/**/*"],
  "exclude": ["__snapshots__/**/*"]
}
