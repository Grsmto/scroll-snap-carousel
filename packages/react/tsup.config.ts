import { defineConfig } from 'tsup';

export default defineConfig({
  entry: ['src/index.ts'], // Adjust the entry point as needed
  format: ['cjs', 'esm'], // CommonJS and ES Module formats
  dts: true, // Generate TypeScript declaration files
  sourcemap: true, // Generate sourcemaps
  // minify: true, // Minify the output
  clean: true, // Clean the output directory before each build
  onSuccess: 'cp ../core/dist/styles.css dist/styles.css'
});
