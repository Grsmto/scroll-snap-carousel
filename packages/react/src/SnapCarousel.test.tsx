import React from 'react';
import { render } from '@testing-library/react';
import { SnapCarousel } from './SnapCarousel';

test('renders without crashing', () => {
  const { container } = render(
    <SnapCarousel>
      {[...Array(5)].map((_, i) => (
        <p key={i}>Some content</p>
      ))}
    </SnapCarousel>
  );
  expect(container.firstChild).toMatchSnapshot();
});
