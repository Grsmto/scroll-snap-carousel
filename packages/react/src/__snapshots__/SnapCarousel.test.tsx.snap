// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders without crashing 1`] = `
<div
  class="snap-carousel-container undefined"
>
  <p>
    Some content
  </p>
  <p>
    Some content
  </p>
  <p>
    Some content
  </p>
  <p>
    Some content
  </p>
  <p>
    Some content
  </p>
</div>
`;
